components/iam/pkg/server/oidc.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package server
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"golang.org/x/oauth2"
+)
+
+type OAuth2Result struct {
+	OAuth2Token *oauth2.Token
+	Redirect    string
+}
+
+type OIDCClientConfig struct {
+	OAuth2Config *oauth2.Config
+	OIDCConfig   *oidc.Config
+}
+
+type OIDCClient struct {
+	Verifier *oidc.IDTokenVerifier
+	Config   *OIDCClientConfig
+}
+
+func (client *OIDCClient) initiateFlow() http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		log.Debug("initiateFlow")
+
+		state, err := randString(32)
+		if err != nil {
+			http.Error(rw, "failed to initiate", http.StatusInternalServerError)
+			return
+		}
+
+		nonce, err := randString(32)
+		if err != nil {
+			http.Error(rw, "failed to initiate", http.StatusInternalServerError)
+			return
+		}
+
+		setCallbackCookie(rw, r, stateCookieName, state)
+		setCallbackCookie(rw, r, nonceCookieName, nonce)
```

Comment:
Anti-pattern here. It's useful to have helpers which create objects for you, but try to not modify other objects inside sub-functions. It makes debugging it rather hard.
```suggestion
		http.SetCookie(rw, newCallbackCookie(...))
		http.SetCookie(rw, newCallbackCookie(...))
```